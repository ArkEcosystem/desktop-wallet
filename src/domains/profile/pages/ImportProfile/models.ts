export type ImportFile = {
	content: string;
	extension: string;
	name: string;
};
