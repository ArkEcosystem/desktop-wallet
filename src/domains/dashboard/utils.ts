import { Profile, ProfileSetting } from "@arkecosystem/platform-sdk-profiles";
import { uniq } from "@arkecosystem/utils";

export const enableNetworkInDashboardFilters = (profile: Profile, networkId: string) => {
	const config: any = profile.settings().get(ProfileSetting.DashboardConfiguration) || { selectedNetworkIds: [] };
	config.selectedNetworkIds.push(networkId);
	profile.settings().set(ProfileSetting.DashboardConfiguration, {
		...config,
		selectedNetworkIds: uniq(config.selectedNetworkIds),
	});
};
