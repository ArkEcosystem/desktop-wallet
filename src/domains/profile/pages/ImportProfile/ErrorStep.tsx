import { Button } from "app/components/Button";
import { Header } from "app/components/Header";
import { FilePreview } from "domains/profile/components/FilePreview";
import React from "react";
import { useTranslation } from "react-i18next";

import { ImportFile } from "./models";

interface ImportErrorProperties {
	file: ImportFile;
	onRetry?: () => void;
	onBack?: () => void;
}

export const ImportError = ({ file, onBack, onRetry }: ImportErrorProperties) => {
	const { t } = useTranslation();

	return (
		<div className="mx-auto max-w-xl" data-testid="ImportError">
			<Header title={t("PROFILE.IMPORT.TITLE")} subtitle={t("PROFILE.IMPORT.ERROR_STEP.DESCRIPTION")} />

			<div className="mt-8">
				<FilePreview file={file} variant="danger" />
			</div>

			<div className="flex justify-end mt-8 space-x-3">
				<Button data-testid="ImportError__back" variant="secondary" onClick={onBack}>
					{t("COMMON.BACK_TO_HOME")}
				</Button>
				<Button data-testid="ImportError__retry" onClick={onRetry}>
					{t("COMMON.RETRY")}
				</Button>
			</div>
		</div>
	);
};
