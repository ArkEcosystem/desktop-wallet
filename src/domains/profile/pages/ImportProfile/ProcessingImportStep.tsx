import { Environment, Profile } from "@arkecosystem/platform-sdk-profiles";
import { Header } from "app/components/Header";
import { FilePreview } from "domains/profile/components/FilePreview";
import { PasswordModal } from "domains/profile/components/PasswordModal";
import { useProfileImport } from "domains/profile/hooks/use-profile-import";
import React, { useEffect, useState } from "react";
import { useTranslation } from "react-i18next";

import { ImportFile } from "./models";

type ProcessingImportProps = {
	env: Environment;
	file: ImportFile;
	password?: string;
	onSuccess?: (profile: any) => void;
	onBack?: () => void;
	onPasswordChange?: (password?: string) => void;
	onError?: (message: string) => void;
	shouldRequestPassword?: boolean;
};

export const ProcessingImport = ({
	file,
	env,
	onError,
	onBack,
	onPasswordChange,
	password,
	onSuccess,
	shouldRequestPassword = false,
}: ProcessingImportProps) => {
	const { t } = useTranslation();
	const { importProfile } = useProfileImport({ env });
	const [isPasswordModalOpen, setIsPasswordModalOpen] = useState(shouldRequestPassword);
	const [passwordError, setPasswordError] = useState<string>();

	useEffect(() => {
		const runImport = async () => {
			let profile: Profile | undefined;

			try {
				if (isPasswordModalOpen) {
					setIsPasswordModalOpen(false);
				}

				profile = await importProfile({ file, password });
				onSuccess?.(profile);
			} catch (error) {
				if (error.message === "PasswordRequired") {
					setIsPasswordModalOpen(true);
					return;
				}

				if (error.message === "InvalidPassword") {
					setIsPasswordModalOpen(true);
					setPasswordError(t("COMMON.VALIDATION.SUBJECT_INVALID", { subject: t("COMMON.PASSWORD") }));
					return;
				}

				onError?.(error.message);
			}
		};

		if (file) {
			runImport();
		}
	}, [file, env, password]); // eslint-disable-line react-hooks/exhaustive-deps

	return (
		<div className="mx-auto max-w-xl" data-testid="ProcessingImport">
			<Header
				title={t("PROFILE.IMPORT.TITLE")}
				subtitle={t("PROFILE.IMPORT.PROCESSING_IMPORT_STEP.DESCRIPTION", { name: file?.name })}
			/>
			<div className="text-theme-secondary-text">{t("PROFILE.IMPORT.PROCESSING_IMPORT_STEP.PLEASE_WAIT")}</div>

			<div className="mt-8">
				<FilePreview file={file} variant="loading" />
			</div>

			<div className="text-left items-left">
				<PasswordModal
					error={passwordError}
					isOpen={isPasswordModalOpen}
					title={t("PROFILE.IMPORT.PASSWORD_TITLE")}
					description={t("PROFILE.IMPORT.PASSWORD_DESCRIPTION")}
					onSubmit={(enteredPassword) => {
						setPasswordError(undefined);
						setIsPasswordModalOpen(false);
						onPasswordChange?.(enteredPassword);
					}}
					onClose={() => {
						setPasswordError(undefined);
						setIsPasswordModalOpen(false);
						onPasswordChange?.(undefined);
						onBack?.();
					}}
				/>
			</div>
		</div>
	);
};
