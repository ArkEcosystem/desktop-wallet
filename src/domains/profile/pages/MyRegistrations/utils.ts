import { DelegateService, ExtendedTransactionData, ReadWriteWallet } from "@arkecosystem/platform-sdk-profiles";

export const filterEntities = (entities: ExtendedTransactionData[], query: string) => {
	if (!query || query === "") return entities;

	const matchesAssetName = (entity: ExtendedTransactionData, query: string) => {
		const { data }: any = entity.asset();
		return data.name.toLowerCase().includes(query.toLowerCase());
	};

	const matchesAddress = (entity: ExtendedTransactionData, query: string) =>
		entity.wallet().address().toLowerCase().includes(query.toLowerCase());

	const matchesWalletName = (entity: ExtendedTransactionData, query: string) =>
		entity.wallet().alias()?.toLowerCase().includes(query.toLowerCase());

	return entities.filter(
		(e) => matchesAssetName(e, query) || matchesWalletName(e, query) || matchesAddress(e, query),
	);
};

export const filterDelegates = (delegateWallets: ReadWriteWallet[], delegates: DelegateService, query: string) => {
	const matchesUserName = (wallet: ReadWriteWallet, query: string) => {
		const delegateInfo = delegates.findByAddress(wallet.coinId(), wallet.networkId(), wallet.address());
		return delegateInfo.username()?.toLowerCase().includes(query);
	};

	return delegateWallets.filter((e) => matchesUserName(e, query));
};
