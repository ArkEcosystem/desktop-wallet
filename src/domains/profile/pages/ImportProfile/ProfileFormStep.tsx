import { Contracts, Environment, Helpers } from "@arkecosystem/platform-sdk-profiles";
import { Button } from "app/components/Button";
import { Divider } from "app/components/Divider";
import { Form, FormField, FormLabel } from "app/components/Form";
import { Header } from "app/components/Header";
import { InputDefault, InputPassword } from "app/components/Input";
import { ListDivided } from "app/components/ListDivided";
import { Select } from "app/components/SelectDropdown";
import { SelectProfileImage } from "app/components/SelectProfileImage";
import { Toggle } from "app/components/Toggle";
import { useTheme, useValidation } from "app/hooks";
import { ReadableFile } from "app/hooks/use-files";
import { PlatformSdkChoices } from "data";
import { FilePreview } from "domains/profile/components/FilePreview";
import React, { useEffect, useLayoutEffect, useMemo, useState } from "react";
import { useForm } from "react-hook-form";
import { useTranslation } from "react-i18next";
import { setThemeSource } from "utils/electron-utils";

interface CreateProfileFormProperties {
	file?: ReadableFile;
	profile: Contracts.IProfile;
	password?: string;
	env: Environment;
	shouldValidate?: boolean;
	onSubmit?: (profile: Contracts.IProfile) => void;
	onBack?: () => void;
}

const CreateProfileForm = ({
	profile,
	env,
	password,
	onSubmit,
	onBack,
	shouldValidate = false,
}: CreateProfileFormProperties) => {
	const { t } = useTranslation();

	const form = useForm<any>({
		defaultValues: {
			confirmPassword: password,
			currency: profile?.settings().get(Contracts.ProfileSetting.ExchangeCurrency),
			isDarkMode: profile?.settings().get(Contracts.ProfileSetting.Theme) === "dark",
			name: profile?.name() || "",
			password,
		},
		mode: "onChange",
	});

	const { watch, register, formState, setValue, trigger } = form;
	const { name, confirmPassword, isDarkMode, currency } = watch([
		"name",
		"confirmPassword",
		"currency",
		"isDarkMode",
	]);

	const [avatarImage, setAvatarImage] = useState(profile?.avatar());

	const { theme, setTheme } = useTheme();
	const { createProfile, password: passwordValidation } = useValidation();

	const formattedName = name?.trim();

	const isSvg = useMemo(() => avatarImage?.endsWith("</svg>"), [avatarImage]);

	useEffect(() => {
		if (shouldValidate) {
			trigger();
		}
	}, [shouldValidate, trigger]);

	useEffect(() => {
		if (!formattedName && isSvg) {
			setAvatarImage("");
		}
	}, [formattedName, isSvg, setAvatarImage]);

	useLayoutEffect(() => {
		if (theme === "dark") {
			setValue("isDarkMode", true);
		}
	}, []); // eslint-disable-line react-hooks/exhaustive-deps

	useEffect(() => {
		const newTheme = isDarkMode ? "dark" : "light";

		document.body.classList.remove(`theme-${theme}`);
		document.body.classList.add(`theme-${newTheme}`);

		setThemeSource(newTheme);
	}, [isDarkMode, theme]);

	const otherItems = [
		{
			isFloatingLabel: true,
			label: t("SETTINGS.GENERAL.OTHER.DARK_THEME.TITLE"),
			labelAddon: <Toggle ref={register()} name="isDarkMode" />,
			labelClass: "text-xl font-semibold",
			labelDescription: t("SETTINGS.GENERAL.OTHER.DARK_THEME.DESCRIPTION"),
		},
	];

	const handleSubmit = async ({ name, password: enteredPassword, currency, isDarkMode }: any) => {
		profile = profile || env.profiles().create(name.trim());

		env.profiles().push(profile);
		await env.profiles().restore(profile, password);

		profile.settings().set(Contracts.ProfileSetting.Name, name);
		profile.settings().set(Contracts.ProfileSetting.Theme, isDarkMode ? "dark" : "light");
		profile.settings().set(Contracts.ProfileSetting.Avatar, avatarImage);
		profile.settings().set(Contracts.ProfileSetting.ExchangeCurrency, currency);

		if (enteredPassword || password) {
			profile.auth().setPassword(enteredPassword || password);
		}

		onSubmit?.(profile);
	};

	const showPasswordFields = !profile?.usesPassword();

	return (
		<div>
			<Form context={form} onSubmit={handleSubmit} data-testid="CreateProfile__form">
				<div className="mt-2">
					<div className="relative mt-8 space-y-5">
						<div className="flex justify-between items-end -mt-4">
							<div className="mr-6 w-full">
								<FormField name="name">
									<FormLabel label={t("SETTINGS.GENERAL.PERSONAL.NAME")} />
									<InputDefault
										ref={register(createProfile.name())}
										onBlur={() => {
											/* istanbul ignore else */
											if (!avatarImage?.length || isSvg) {
												setAvatarImage(Helpers.Avatar.make(formattedName));
											}
										}}
									/>
								</FormField>
							</div>

							<SelectProfileImage
								value={avatarImage}
								name={formattedName}
								showLabel={false}
								onSelect={setAvatarImage}
							/>
						</div>

						{showPasswordFields && (
							<>
								<FormField name="password">
									<FormLabel label={t("SETTINGS.GENERAL.PERSONAL.PASSWORD")} optional />
									<InputPassword
										ref={register(passwordValidation.password())}
										onChange={() => {
											if (confirmPassword) {
												trigger("confirmPassword");
											}
										}}
									/>
								</FormField>

								<FormField name="confirmPassword">
									<FormLabel
										label={t("SETTINGS.GENERAL.PERSONAL.CONFIRM_PASSWORD")}
										optional={!watch("password")}
									/>
									<InputPassword
										ref={register(passwordValidation.confirmOptionalPassword(watch("password")!))}
									/>
								</FormField>
							</>
						)}

						<FormField name="currency">
							<FormLabel label={t("SETTINGS.GENERAL.PERSONAL.CURRENCY")} />
							<Select
								defaultValue={currency}
								placeholder={t("COMMON.SELECT_OPTION", {
									option: t("SETTINGS.GENERAL.PERSONAL.CURRENCY"),
								})}
								ref={register(createProfile.currency())}
								options={PlatformSdkChoices.currencies}
								onChange={(currency: any) =>
									setValue("currency", currency?.value, {
										shouldDirty: true,
										shouldValidate: true,
									})
								}
							/>
						</FormField>
					</div>

					<div className="pb-4 mt-8">
						<ListDivided items={otherItems} />
					</div>
				</div>

				<Divider />

				<div className="flex justify-end pt-4 space-x-3">
					<Button
						variant="secondary"
						onClick={() => {
							onBack?.();
							// to prevent changing theme by component
							setTimeout(() => setTheme("system"), 0);
						}}
						data-testid="CreateProfile__back-button"
					>
						{t("COMMON.BACK")}
					</Button>

					<Button disabled={!formState.isValid} type="submit" data-testid="CreateProfile__submit-button">
						{t("COMMON.CREATE")}
					</Button>
				</div>
			</Form>
		</div>
	);
};

export const ImportProfileForm = ({
	profile,
	env,
	onSubmit,
	onBack,
	file,
	password,
	shouldValidate,
}: CreateProfileFormProperties) => {
	const { t } = useTranslation();

	return (
		<div className="mx-auto max-w-xl">
			<Header title={t("PROFILE.IMPORT.TITLE")} subtitle={t("PROFILE.IMPORT.FORM_STEP.DESCRIPTION")} />

			<div className="flex flex-col px-10 pt-8 pb-10 mt-10 rounded-lg border bg-theme-background border-theme-secondary-300 dark:border-theme-secondary-800">
				<div>
					<FilePreview file={file} variant="success" useBorders={false} />
				</div>

				<Divider />

				<CreateProfileForm
					shouldValidate={shouldValidate}
					profile={profile}
					env={env}
					onSubmit={onSubmit}
					onBack={onBack}
					password={password}
				/>
			</div>
		</div>
	);
};
