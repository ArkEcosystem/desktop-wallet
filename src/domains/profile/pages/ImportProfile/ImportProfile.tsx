import { Contracts } from "@arkecosystem/platform-sdk-profiles";
import { Page, Section } from "app/components/Layout";
import { StepIndicator } from "app/components/StepIndicator";
import { TabPanel, Tabs } from "app/components/Tabs";
import { useEnvironmentContext } from "app/contexts";
import { ReadableFile } from "app/hooks/use-files";
import { ImportError } from "domains/profile/pages/ImportProfile/ErrorStep";
import { ProcessingImport } from "domains/profile/pages/ImportProfile/ProcessingImportStep";
import { ImportProfileForm } from "domains/profile/pages/ImportProfile/ProfileFormStep";
import { SelectFileStep } from "domains/profile/pages/ImportProfile/SelectFileStep";
import React, { useState } from "react";
import { useTranslation } from "react-i18next";
import { useHistory } from "react-router-dom";

export const ImportProfile = () => {
	const { env, persist } = useEnvironmentContext();
	const { t } = useTranslation();
	const history = useHistory();

	const [activeTab, setActiveTab] = useState(1);
	const [fileFormat, setFileFormat] = useState(".dwe");
	const [selectedFile, setSelectedFile] = useState<ReadableFile>();
	const [password, setPassword] = useState<string>();
	const [profile, setProfile] = useState<Contracts.IProfile>();

	const handleSelectedFile = (file: ReadableFile) => {
		setSelectedFile(file);
		setActiveTab(2);
	};

	const handleImportError = () => {
		setActiveTab(10);
	};

	const handleProfileSave = (importedProfile: Contracts.IProfile) => {
		persist();
		history.push("/");
	};

	return (
		<Page navbarVariant="logo-only" title={t("COMMON.DESKTOP_WALLET")}>
			<Section className="flex flex-col">
				<div className="mx-auto max-w-xl">
					<StepIndicator size={3} activeIndex={activeTab} />

					<Tabs activeId={activeTab} className="mt-8">
						<TabPanel tabId={1}>
							<SelectFileStep
								fileFormat={fileFormat}
								onFileFormatChange={setFileFormat}
								onSelect={handleSelectedFile}
								onBack={() => history.push("/")}
							/>
						</TabPanel>

						<TabPanel tabId={2}>
							{selectedFile && (
								<ProcessingImport
									env={env}
									password={password}
									file={selectedFile}
									onSuccess={(profile) => {
										setProfile(profile);
										setActiveTab(3);
									}}
									onPasswordChange={setPassword}
									onBack={() => setActiveTab(1)}
									onError={handleImportError}
								/>
							)}
						</TabPanel>

						<TabPanel tabId={3}>
							{profile && (
								<ImportProfileForm
									file={selectedFile}
									env={env}
									profile={profile}
									password={password}
									shouldValidate
									onSubmit={handleProfileSave}
									onBack={() => {
										setPassword(undefined);
										setActiveTab(1);
									}}
								/>
							)}
						</TabPanel>

						<TabPanel tabId={10}>
							{selectedFile && (
								<ImportError
									file={selectedFile}
									onRetry={() => handleSelectedFile(selectedFile)}
									onBack={() => history.push("/")}
								/>
							)}
						</TabPanel>
					</Tabs>
				</div>
			</Section>
		</Page>
	);
};
