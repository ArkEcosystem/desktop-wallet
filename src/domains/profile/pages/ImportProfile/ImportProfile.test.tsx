/* eslint-disable @typescript-eslint/require-await */
import { EnvironmentProvider } from "app/contexts";
import { ImportProfile } from "domains/profile/pages/ImportProfile/ImportProfile";
import fs from "fs";
import { createMemoryHistory } from "history";
import React from "react";
import { act, env, fireEvent, renderWithRouter, waitFor } from "utils/testing-library";

const passwordProtectedDwe = fs.readFileSync("src/tests/fixtures/profile/import/password-protected-profile.dwe");
const corruptedDwe = fs.readFileSync("src/tests/fixtures/profile/import/corrupted-profile.dwe");
const history = createMemoryHistory();

describe("ImportProfile", () => {
	let consoleSpy: jest.SpyInstance;
	let fsMock: jest.SpyInstance;

	beforeAll(() => {
		consoleSpy = jest.spyOn(console, "log").mockImplementation(() => undefined);
		console.log("");
		fsMock = jest.spyOn(fs, "readFileSync").mockReturnValue(passwordProtectedDwe);
	});

	afterAll(() => {
		consoleSpy.mockRestore();
		fsMock.mockRestore();
	});

	it("should render first step", async () => {
		const history = createMemoryHistory();
		history.push(`/profiles/import`);

		const { container, getByTestId } = renderWithRouter(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
		);

		expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it("should go back", async () => {
		history.push(`/profiles/import`);
		const historyMock = jest.spyOn(history, "push").mockReturnValue();

		const { getByTestId } = renderWithRouter(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument();
		expect(getByTestId("SelectFileStep__back")).toBeInTheDocument();

		act(() => {
			fireEvent.click(getByTestId("SelectFileStep__back"));
		});

		await waitFor(() => expect(historyMock).toHaveBeenCalled());
		historyMock.mockRestore();
	});

	it("should change file format", async () => {
		history.push(`/profiles/import`);

		const { getByTestId } = renderWithRouter(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument();

		act(() => {
			fireEvent.click(getByTestId("SelectFileStep__change-file"));
		});

		await waitFor(() => expect(() => getByTestId("SelectFileStep__change-file")).toThrow());
	});

	it("should select file and go to step 2", async () => {
		history.push(`/profiles/import`);

		const { getByTestId } = renderWithRouter(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument();
		expect(getByTestId("SelectFileStep__back")).toBeInTheDocument();

		act(() => {
			fireEvent.drop(getByTestId("SelectFile__browse-files"), {
				dataTransfer: {
					files: [{ name: "profile-export.dwe", path: "path/to/sample-export.dwe" }],
				},
			});
		});

		await waitFor(() => expect(getByTestId("ProcessingImport")).toBeInTheDocument());
	});

	it("should request and set password for importing password protected profile", async () => {
		history.push(`/profiles/import`);

		const { getByTestId, findByTestId } = renderWithRouter(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument();
		expect(getByTestId("SelectFileStep__back")).toBeInTheDocument();

		act(() => {
			fireEvent.drop(getByTestId("SelectFile__browse-files"), {
				dataTransfer: {
					files: [{ name: "profile-export.dwe", path: "path/to/sample-export.dwe" }],
				},
			});
		});

		await waitFor(() => expect(getByTestId("ProcessingImport")).toBeInTheDocument());
		await waitFor(() => expect(getByTestId("modal__inner")).toBeInTheDocument());

		await act(async () => {
			fireEvent.input(getByTestId("PasswordModal__input"), { target: { value: "testtest" } });
		});

		// wait for formState.isValid to be updated
		await findByTestId("PasswordModal__submit-button");

		act(() => {
			fireEvent.click(getByTestId("PasswordModal__submit-button"));
		});

		await waitFor(() => expect(getByTestId("ProcessingImport")).toBeInTheDocument());
	});

	it("should close password modal and go back to select file", async () => {
		history.push(`/profiles/import`);

		const { getByTestId } = renderWithRouter(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument();
		expect(getByTestId("SelectFileStep__back")).toBeInTheDocument();

		act(() => {
			fireEvent.drop(getByTestId("SelectFile__browse-files"), {
				dataTransfer: {
					files: [{ name: "profile-export.dwe", path: "path/to/sample-export.dwe" }],
				},
			});
		});

		await waitFor(() => expect(getByTestId("ProcessingImport")).toBeInTheDocument());
		await waitFor(() => expect(getByTestId("modal__inner")).toBeInTheDocument());

		act(() => {
			fireEvent.click(getByTestId("modal__close-btn"));
		});

		await waitFor(() => expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument());
	});

	it("should succesfully import profile and return to home screen", async () => {
		history.push(`/profiles/import`);
		const historyMock = jest.spyOn(history, "push").mockReturnValue();
		const fsReadFileMock = jest.spyOn(fs, "readFileSync").mockReturnValue(passwordProtectedDwe);

		const { getByTestId } = renderWithRouter(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument();
		expect(getByTestId("SelectFileStep__back")).toBeInTheDocument();

		act(() => {
			fireEvent.drop(getByTestId("SelectFile__browse-files"), {
				dataTransfer: {
					files: [{ name: "profile-export.dwe", path: "path/to/sample-export.dwe" }],
				},
			});
		});

		await waitFor(() => expect(getByTestId("ProcessingImport")).toBeInTheDocument());
		await waitFor(() => expect(getByTestId("modal__inner")).toBeInTheDocument());

		await act(async () => {
			fireEvent.input(getByTestId("PasswordModal__input"), { target: { value: "testtest" } });
		});

		act(() => {
			fireEvent.click(getByTestId("PasswordModal__submit-button"));
		});

		await waitFor(() => expect(getByTestId("ProcessingImport")).toBeInTheDocument());
		await waitFor(() => expect(getByTestId("CreateProfile__form")).toBeInTheDocument());

		act(() => {
			fireEvent.click(getByTestId("CreateProfile__submit-button"));
		});

		await waitFor(() => expect(historyMock).toHaveBeenCalledWith("/"));
	});

	it("should go to step 3 and back", async () => {
		history.push(`/profiles/import`);

		const { getByTestId } = renderWithRouter(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument();
		expect(getByTestId("SelectFileStep__back")).toBeInTheDocument();

		act(() => {
			fireEvent.drop(getByTestId("SelectFile__browse-files"), {
				dataTransfer: {
					files: [{ name: "profile-export.dwe", path: "path/to/sample-export.dwe" }],
				},
			});
		});

		await waitFor(() => expect(getByTestId("ProcessingImport")).toBeInTheDocument());
		await waitFor(() => expect(getByTestId("modal__inner")).toBeInTheDocument());

		await act(async () => {
			fireEvent.input(getByTestId("PasswordModal__input"), { target: { value: "testtest" } });
		});

		act(() => {
			fireEvent.click(getByTestId("PasswordModal__submit-button"));
		});

		await waitFor(() => expect(getByTestId("ProcessingImport")).toBeInTheDocument());
		await waitFor(() => expect(getByTestId("CreateProfile__form")).toBeInTheDocument());

		act(() => {
			fireEvent.click(getByTestId("CreateProfile__back-button"));
		});

		await waitFor(() => expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument());
	});

	it("should fail profile import and show error step", async () => {
		history.push(`/profiles/import`);
		const corruptedDweMock = jest.spyOn(fs, "readFileSync").mockReturnValue(corruptedDwe);

		const { getByTestId } = renderWithRouter(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument();
		expect(getByTestId("SelectFileStep__back")).toBeInTheDocument();

		act(() => {
			fireEvent.drop(getByTestId("SelectFile__browse-files"), {
				dataTransfer: {
					files: [{ name: "profile-export.dwe", path: "path/to/sample-export.dwe" }],
				},
			});
		});

		await waitFor(() => expect(getByTestId("ProcessingImport")).toBeInTheDocument());
		await waitFor(() => expect(getByTestId("modal__inner")).toBeInTheDocument());

		await act(async () => {
			fireEvent.input(getByTestId("PasswordModal__input"), { target: { value: "wrong password" } });
		});

		act(() => {
			fireEvent.click(getByTestId("PasswordModal__submit-button"));
		});

		await waitFor(() => expect(getByTestId("ImportError")).toBeInTheDocument());

		corruptedDweMock.mockRestore();
	});

	it("should fail profile import and retry", async () => {
		history.push(`/profiles/import`);
		const corruptedDweMock = jest.spyOn(fs, "readFileSync").mockReturnValue(corruptedDwe);

		const { getByTestId } = renderWithRouter(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument();
		expect(getByTestId("SelectFileStep__back")).toBeInTheDocument();

		act(() => {
			fireEvent.drop(getByTestId("SelectFile__browse-files"), {
				dataTransfer: {
					files: [{ name: "profile-export.dwe", path: "path/to/sample-export.dwe" }],
				},
			});
		});

		await waitFor(() => expect(getByTestId("ProcessingImport")).toBeInTheDocument());
		await waitFor(() => expect(getByTestId("modal__inner")).toBeInTheDocument());

		await act(async () => {
			fireEvent.input(getByTestId("PasswordModal__input"), { target: { value: "wrong password" } });
		});

		act(() => {
			fireEvent.click(getByTestId("PasswordModal__submit-button"));
		});

		await waitFor(() => expect(getByTestId("ImportError")).toBeInTheDocument());

		act(() => {
			fireEvent.click(getByTestId("ImportError__retry"));
		});

		await waitFor(() => expect(getByTestId("ImportError")).toBeInTheDocument());
		corruptedDweMock.mockRestore();
	});

	it("should fail profile import and go back to home screen", async () => {
		history.push(`/profiles/import`);
		const corruptedDweMock = jest.spyOn(fs, "readFileSync").mockReturnValue(corruptedDwe);
		const historyMock = jest.spyOn(history, "push").mockReturnValue();

		const { getByTestId } = renderWithRouter(
			<EnvironmentProvider env={env}>
				<ImportProfile />
			</EnvironmentProvider>,
			{ history },
		);

		expect(getByTestId("SelectFileStep__change-file")).toBeInTheDocument();
		expect(getByTestId("SelectFileStep__back")).toBeInTheDocument();

		act(() => {
			fireEvent.drop(getByTestId("SelectFile__browse-files"), {
				dataTransfer: {
					files: [{ name: "profile-export.dwe", path: "path/to/sample-export.dwe" }],
				},
			});
		});

		await waitFor(() => expect(getByTestId("ProcessingImport")).toBeInTheDocument());
		await waitFor(() => expect(getByTestId("modal__inner")).toBeInTheDocument());

		await act(async () => {
			fireEvent.input(getByTestId("PasswordModal__input"), { target: { value: "wrong password" } });
		});

		act(() => {
			fireEvent.click(getByTestId("PasswordModal__submit-button"));
		});

		await waitFor(() => expect(getByTestId("ImportError")).toBeInTheDocument());

		act(() => {
			fireEvent.click(getByTestId("ImportError__back"));
		});

		await waitFor(() => expect(getByTestId("ImportError")).toBeInTheDocument());
		await waitFor(() => expect(historyMock).toHaveBeenCalledWith("/"));

		historyMock.mockRestore();
		corruptedDweMock.mockRestore();
	});
});
