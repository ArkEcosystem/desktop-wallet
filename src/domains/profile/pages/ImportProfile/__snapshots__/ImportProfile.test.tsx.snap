// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportProfile should render first step 1`] = `
<div>
  <div
    class="flex relative flex-col min-h-screen"
  >
    <nav
      aria-labelledby="main menu"
      class="NavigationBar__NavWrapper-sc-1qsoskw-0 noNGK"
    >
      <div
        class="flex relative h-21"
      >
        <div
          class="flex flex-1 px-8 ml-12"
        >
          <div
            class="flex items-center my-auto"
          >
            <div
              class="NavigationBar__LogoContainer-sc-1qsoskw-2 tulfg"
            >
              <svg
                width="44"
              >
                ark-logo.svg
              </svg>
            </div>
            <span
              class="text-2xl font-bold"
            >
              Desktop Wallet
            </span>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="flex flex-col flex-1"
    >
      <div
        class="Section__SectionWrapper-sc-8cjvre-0 kJTPww flex flex-col"
      >
        <div
          class="container px-10 mx-auto"
        >
          <div
            class="mx-auto max-w-xl"
          >
            <ul
              class="StepIndicator__StepWrapper-sc-15t41is-1 iToWWb"
            >
              <li
                class="StepIndicator__StepStyled-sc-15t41is-0 dATCfl"
              />
              <li
                class="StepIndicator__StepStyled-sc-15t41is-0 khoCaB"
              />
              <li
                class="StepIndicator__StepStyled-sc-15t41is-0 khoCaB"
              />
            </ul>
            <div
              class="mt-8"
            >
              <div
                data-testid="tab-pabel__active-panel"
              >
                <div
                  class="mx-auto max-w-xl"
                >
                  <div
                    class="flex justify-between items-end bg-theme-background"
                  >
                    <div
                      class="space-y-4"
                    >
                      <h1
                        class="mb-0 text-4xl"
                        data-testid="header__title"
                      >
                        Import Profile
                      </h1>
                      <div
                        class="text-theme-secondary-text"
                        data-testid="header__subtitle"
                      >
                        Select a .dwe file with your Profile and related properties to start the import process.
                      </div>
                    </div>
                  </div>
                  <div
                    class="h-52 border-2 border-dashed rounded-lg mt-8 border-theme-secondary-300 dark:border-theme-secondary-800 flex flex-col items-center justify-center"
                    data-testid="SelectFile"
                  >
                    <div
                      class="sc-bdfBwQ fTnhmh"
                      height="40"
                      width="40"
                    >
                      <svg>
                        dwe-file.svg
                      </svg>
                    </div>
                    <div
                      class="mt-8"
                    >
                      <span
                        class="font-semibold"
                      >
                        Drag & Drop or
                         
                      </span>
                      <span
                        class="font-semibold cursor-pointer link"
                        data-testid="SelectFile__browse-files"
                        title="Click here to upload"
                      >
                        Browse Files
                      </span>
                    </div>
                    <div
                      class="font-semibold text-theme-secondary-500 text-sm mt-2"
                    >
                      Supported format is .dwe
                    </div>
                  </div>
                  <p
                    class="text-sm text-theme-secondary-text md:text-base mt-6 text-center"
                  >
                    <span>
                      Importing from Desktop Wallet v2?
                       
                    </span>
                    <span
                      class="font-semibold cursor-pointer link"
                      data-testid="SelectFileStep__change-file"
                      title="Click here."
                    >
                      Click here.
                    </span>
                  </p>
                  <div
                    class="flex justify-end mt-8 space-x-3"
                  >
                    <button
                      class="Button__StyledButton-mdtvtf-0 iUSBMV"
                      data-testid="SelectFileStep__back"
                      type="button"
                    >
                      <div
                        class="flex relative items-center space-x-2"
                      >
                        Back
                      </div>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
